### PR TITLE
Mint-005 Proposal

### PR DESCRIPTION
**Pull request by Rob1Ham**
This PR introduces a Miniscript Template for 3 Key Joint Custody Model. We have been internally referring it to Mint-005 as we've had other prior versions we have been testing with, but did not move forward with full implementation.

This is a miniscript template, written for P2WSH that leverages thresholds of thresholds to have a distinction between Principals (holders/owners of bitcoin) and Agents (service providers helping secure bitcoin). Each have 3 keys, and with the utilization of epoch absolute timelocks, have multiple contingency spending conditions which become available over time.

### CODE REVIEW

**Comment by apoelstra** [comment]
Happy to give you number 5 for this proposal. Before I look at it more carefully, can you remove the first two commits (which cancel each other out) and also remove the trailing whitespace on two lines in 6fc11e7346fa4489877e65f93f0d0e40f7ca7eb7?

**Comment by Rob1Ham** [comment]
Cleaned up the commit history and the trailing whitespace.

File: mint-005.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
+# mint-005
+
+## 3 Key Joint Custody
+
+## Motivation
+
+To provide a custody arrangement in which an owner of Bitcoin (principle) is able to secure bitcoin by working with an Agent. Unlike existing collaborative custody models up to this point, **where bitcoin keys WITHIN a multisig threshold are shared between principles and agents**, a joint custody model by default requires a threshold of keys by both the Agent & Principle for movement of funds.
```
Comment (apoelstra) [inline mint-005.md]: In 7bebe7efd503756903ccb9efc65e1e681a5ebb4e:

`principle` should be `principal` throughout, and (I think) capitalized.

Capitalization of `bitcoin` should be consistent (I think lowercase).

`Primary Agent` should be consistently capitalized throughout.

File: mint-005.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+Finally, in the event the principle no longer wishes to work with the agent, say after a contract expires, the custody defers to a set of recovery keys, which can be held either by the principle, or their own delegated managers of the recovery keys. As a result, when enough time has passed, the Principle is able to move Bitcoin unilaterally without having the Agent sign key material.
+
+### More on Timelock Values
+
+There are three timelocks used for this MinT:
+
+1. `smallest_epoch_timestamp` - The smallest epoch timestamp timelock enables a "Asset Recovery" period such that only one of the principle keys is required to sign.
+
+2. `between_epoch_timestamp` - The epoch timestamp value in between the smallest and largest epoch timestamp enables a "Emergency Recover Path". In the event the Principle has lost all of their keys, the Primary agent and Secondary Agent can work together to recover the Bitcoin in the Joint Custody vault.
```
Comment (apoelstra) [inline mint-005.md]: In 7bebe7efd503756903ccb9efc65e1e681a5ebb4e:

I think "Emergency Recovery Path" reads better than "Emergency Recover Path".

File: mint-005.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+| Scenario 24       |      | ✅    | ✅    | ✅                                  |
+
+
+---
+
+# Example Miniscript Output Descriptor
+
+For this example, the `smallest_epoch_timestamp` is: 1672531200 (Jan 1 2023, midnight gmt), the `between_epoch_timestamp` is: 1673740800 and `largest_epoch_timestamp` is: 1675209600 (Feb 1 2023, midnight gmt)
+
+- MINT-005 Output Descriptor:
```
Comment (apoelstra) [inline mint-005.md]: In 7bebe7efd503756903ccb9efc65e1e681a5ebb4e:

trailing whitespace

File: mint-005.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+
+# Example Miniscript Output Descriptor
+
+For this example, the `smallest_epoch_timestamp` is: 1672531200 (Jan 1 2023, midnight gmt), the `between_epoch_timestamp` is: 1673740800 and `largest_epoch_timestamp` is: 1675209600 (Feb 1 2023, midnight gmt)
+
+- MINT-005 Output Descriptor: 
+<code>wsh(andor(multi(2,$PAK_1$,$PAK_2$,$PAK_3$),or_i(and_v(v:pkh($SAK$),after(`between_epoch_timestamp`)),thresh(2,pk($PK_1$),s:pk($PK_2$),s:pk($PK_3$),snl:after(`smallest_epoch_timestamp`))),and_v(v:thresh(2,pkh($RK_1$),a:pkh($RK_2$),a:pkh($RK_3$)),after(`larget_epoch_timestamp`))))</code>
+
+
+- Source Policy (FOR REFERENCE PURPOSES ONLY):
```
Comment (apoelstra) [inline mint-005.md:137]: In 7bebe7efd503756903ccb9efc65e1e681a5ebb4e:

trailing whitespace

File: mint-005.md [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+
+## Layer 1 Example Spend
+
+Signed by: $PK_1$, $PK_2$, $PAK_1$, $PAK_2$
+
+[Reference Testnet
+Transaction](https://mempool.space/signet/tx/2836d6af6b5c4bb01e926391f64771fb333193676040b24d4236ba0bb89a7008)
+
+## Layer 2 Example Spend
```
Comment (apoelstra) [inline mint-005.md]: In 7bebe7efd503756903ccb9efc65e1e681a5ebb4e:

trailing whitespace

**Comment by apoelstra** [comment]
Done reviewing 7bebe7efd503756903ccb9efc65e1e681a5ebb4e. This looks great. It's a little complicated and hard to follow but the separation into layers really helps. I would recommend reducing the size of the key graphics and maybe having different colors for the PK, PAK, and RKs.

I nitted a few grammar/formatting things. Should also run a spellchecker. Otherwise ACK.

File: mint-005.md (hunk 2 above)
Comment (Rob1Ham) [reply mint-005.md]: Updated

File: mint-005.md (hunk 1 above)
Comment (Rob1Ham) [reply mint-005.md]: Updated

File: mint-005.md (hunk 3 above)
Comment (Rob1Ham) [reply mint-005.md]: Updated

File: mint-005.md (hunk 5 above)
Comment (Rob1Ham) [reply mint-005.md]: Updated

**Comment by Rob1Ham** [comment]
I've implemented all of the suggested edits, as well as adding more detail with additional color keys and more context within the tables to make it more straight forward to read in line.

**Review by apoelstra — APPROVED** [review]
ACK beb16afb7f3bd1e23fb2ff92a34704dee9e95289

**Comment by apoelstra** [comment]
Actually, there continues to be trailing whitespace on two lines :P.

**Comment by apoelstra** [comment]
But otherwise this looks great. The new colors help a lot!

**Comment by Rob1Ham** [comment]
Ok, whitespaces fully cleaned up now!

**Review by apoelstra — APPROVED** [review]
ACK 52ba4a0b89a4c5dcd983ce2f5b1e16f3cb128031
